src/viur/core/bones/uri.py

```diff
@@ -0,0 +1,204 @@
+                return [range(PORT_MIN, PORT_MAX + 1)]
+
+            elif "," in accepted_ports:  # list of ranges, values
+                return cls._build_accepted_ports([
+                    value.strip() for value in accepted_ports.split(",")
+                ])
+
+            elif "-" in accepted_ports:  # range of ports
+                start, end = accepted_ports.split("-", 1)
+                start = int(start)
+                end = int(end)
+                if start > end:
+                    raise ValueError("Start value must be less than end value")
+
+                if start < PORT_MIN:
+                    raise ValueError("Start value must be greater than zero")
+
+                if end > PORT_MAX:
+                    raise ValueError(f"End value must be less or equal than {PORT_MAX}")
+
+                return [range(start, end + 1)]
+
+            else:
+                port = int(accepted_ports)
+                return [range(port, port + 1)]
+
+        elif isinstance(accepted_ports, int):
+            if accepted_ports < PORT_MIN:
+                raise ValueError("Port value must be greater than zero")
+
+            if accepted_ports > PORT_MAX:
+                raise ValueError(f"Port value must be less or equal than {PORT_MAX}")
+
+            return [range(accepted_ports, accepted_ports + 1)]
+
+        elif isinstance(accepted_ports, Iterable):
+            accepted_ports_value = []
+            for accepted_port in accepted_ports:
+                accepted_ports_value.extend(UriBone._build_accepted_ports(accepted_port))
+            return accepted_ports_value
+
+        raise ValueError("accepted_ports must be a iterable or an integer or string")
+
+    def isInvalid(self, value) -> str | None:
+        try:
+            parsed_url = urlparse(value)
+        except ValueError:
+            return "Can't parse URL"
+
+        if not self.local_path_allowed and parsed_url.scheme == "":
+            return f"""No protocol specified"""
+
+        if self.accepted_ports:
+            if not any(parsed_url.port in rng for rng in self.accepted_ports):
+                return f""""{parsed_url.port}" not in the accepted ports."""
+
+        if self.accepted_protocols:
+            for protocol in self.accepted_protocols:
+                if fnmatch.fnmatch(parsed_url.scheme, protocol):
+                    break
+            else:
+                return f""""{parsed_url.scheme}" not in the accepted protocols."""
+
+        if self.domain_allowed_list is not None:
+            for domain in self.domain_allowed_list:
+                if fnmatch.fnmatch(value, domain):
+                    break
+            else:
+                return f"""Url is not in the domain allowed list."""
+
+        if self.domain_disallowed_list is not None:
+            for domain in self.domain_disallowed_list:
+                if fnmatch.fnmatch(value, domain):
+                    return f"""Url is in the domain disallowed list."""
```

Comment:
These tests are applied to the full value.

When I read `domain_allow_list` in the docstring and the parameter signature, I suggest that `["mausbrand.de"]` suffices for a domain allow list, like so:
```py
    url = UriBone(
        accepted_protocols=("https", "http"),
        domain_allowed_list=["mausbrand.de"],
    )
```
but instead, I must provide `["https://*.mausbrand.de/*", "https://mausbrand.de/*", "https://mausbrand.de", "http://*.mausbrand.de/*", "http://mausbrand.de/*", "http://mausbrand.de"]` to match all domains like subdomains, tld-only and all protocols (and this still doesn't match all cases).

This isn't cool. Maybe the bone should also accept regex here, but this is stage 2.

Reply:
You can provide it like 

    url = UriBone(
        accepted_protocols=("https", "http"),
        domain_allowed_list=["*mausbrand.de"],
    )
for all of this ?

Reply:
> You can provide it like
> 
> ```python
>     url = UriBone(
>         accepted_protocols=("https", "http"),
>         domain_allowed_list=["*mausbrand.de"],
>     )
> ```
> 
> for all of this ?

Sure! But the protocol is not part of the domain, and shall not be handled as so. What if I want to accept only `mausbrand.de`? Then I must provide `*//mausbrand.de`? This notation is ulgy.

Oh right, and given `["*mausbrand.de"]`, I cannot provide a link `https://mausbrand.de/test` because then, the definition must be `["*mausbrand.de*"]`, but this also accepts `mausbrand.decoration` as accepted domain name.

Err no... please correct it so that the domain_allow/disalllow-lists are only matched against the true domain that is given, and not the entire value.

Reply:
Can you implement some testcases based on this example?